arangod/Aql/IResearchViewExecutor.tpp

```diff
@@ -1412,122 +1439,265 @@ bool IResearchViewHeapSortExecutor<ExecutionTraits>::fillBufferInternal(
+  bool gotData = false;
+  auto atMost = ctx.outputRow.numRowsLeft();
+  TRI_ASSERT(this->_indexReadBuffer.empty());
+  this->_isMaterialized = false;
+  auto parallelism = std::min(count, this->_infos.parallelism());
+  this->_indexReadBuffer.reset();
+  std::atomic<size_t> bufferIdx{0};
+  // shortcut for sequential execution.
+  if (parallelism == 1) {
+    TRI_IF_FAILURE("IResearchFeature::failNonParallelQuery") {
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
+    }
+    this->_indexReadBuffer.preAllocateStoredValuesBuffer(
+        atMost, this->_infos.getScoreRegisters().size(),
+        this->_infos.getOutNonMaterializedViewRegs().size());
+    auto& reader = _segmentReaders.front();
+    while (!gotData && (_segmentOffset < count || reader.itr)) {
+      if (!reader.itr) {
+        reader.readerOffset = _segmentOffset++;
+        TRI_ASSERT(reader.readerOffset < count);
+      }
+      reader.atMost = atMost;
+      gotData = readSegment<false>(reader, bufferIdx);
+    }
+    return gotData;
+  }
+  auto const atMostInitial = atMost;
+  // here parallelism can be used or not depending on the
+  // current pipeline demand.
+  auto const& clientCall = ctx.outputRow.getClientCall();
+  auto limit = clientCall.getUnclampedLimit();
+  bool const isUnlimited = limit == aql::AqlCall::Infinity{};
+  TRI_ASSERT(isUnlimited || std::holds_alternative<std::size_t>(limit));
+  if (!isUnlimited) {
+    parallelism = std::min(parallelism,
+                           (std::get<std::size_t>(limit) / atMostInitial) + 1);
+  }
+  // let's be greedy as it is more likely that we are
+  // asked to read some "tail" documents to fill the block
+  // and next time we would need all our parallelism again.
+  auto& readersPool = this->infos().parallelExecutionPool();
+  if (parallelism > (this->_allocatedThreads + 1)) {
+    this->_allocatedThreads += readersPool.allocateThreads(
+        static_cast<int>(parallelism - this->_allocatedThreads - 1));
+    parallelism = this->_allocatedThreads + 1;
+  }
+  atMost = atMostInitial * parallelism;
+
+  std::vector<futures::Future<bool>> results;
+  this->_indexReadBuffer.preAllocateStoredValuesBuffer(
+      atMost, this->_infos.getScoreRegisters().size(),
+      this->_infos.getOutNonMaterializedViewRegs().size());
+  this->_indexReadBuffer.setForParallelAccess(
+      atMost, this->_infos.getScoreRegisters().size(),
+      this->_infos.getOutNonMaterializedViewRegs().size());
+  results.reserve(parallelism - 1);
+  // we must wait for our threads before bailing out
+  // as we most likely will release segments and
+  // "this" will also be invalid.
+  auto cleanupThreads = irs::Finally([&]() noexcept {
+    results.erase(std::remove_if(results.begin(), results.end(),
+                                 [](auto& f) { return !f.valid(); }),
+                  results.end());
+    if (results.empty()) {
+      return;
+    }
+    auto runners = futures::collectAll(results);
+    runners.wait();
+  });
```

Comment:
You can unify this code if you will use absl::Cleanup and Invoke (call cleanup function before dtor)